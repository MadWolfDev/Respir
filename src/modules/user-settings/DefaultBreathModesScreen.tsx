import { Button } from '@mui/material';
import { styled } from '@mui/material/styles';
import { useBreathModeStore } from '../../store/breathModeStore';
import { motion } from 'framer-motion';
import { Dispatch, SetStateAction } from 'react';
import { Display } from './WelcomeScreen';
import { BreathModes } from '../../store/defaultBreathConfigs';
import { BreathModesDisplay } from '../../store/BreathModesDisplay.type';
import { AnimVariant } from '../breath-activity/createBreathAnimationVariants';

export const DefaultBreathModesScreen = ({
  setCurrentDisplay,
}: {
  setCurrentDisplay: Dispatch<SetStateAction<Display>>;
}) => {
  const updateBreathModeSelection = useBreathModeStore(
    (state) => state.updateBreathMode
  );
  const changeBreathMode = (newBreathMode: BreathModes) => {
    updateBreathModeSelection(newBreathMode);
    newBreathMode === 'custom'
      ? setCurrentDisplay(Display.customModeDisplay)
      : setCurrentDisplay(Display.empty);
  };

  const ButtonExitAnim: ExitAnim = {
    y: '0em',
    transition: { duration: 0.75, delay: 0, ease: 'anticipate' },
  };

  const ButtonEnterAnimVariant: AnimVariant = {
    initial: { y: '5em' },
    animate: {
      y: '0em',
      transition: { duration: 0.75, ease: 'backOut' },
    },
  };

  const ButtonsExitAnims: ExitAnim[] = new Array(5);
  ButtonsExitAnims.fill(ButtonExitAnim);
  const ButtonsEnterAnimVariants: AnimVariant[] = new Array(5);
  ButtonsEnterAnimVariants.fill(ButtonEnterAnimVariant);

  for (let i = 0; i < 5; i++) {
    ButtonsExitAnims[i] = {
      ...ButtonsExitAnims[i],
      y: `${2.5}em`,
      transition: { ...ButtonsExitAnims[i].transition, delay: i * 0.05 },
    };

    ButtonsEnterAnimVariants[i] = {
      ...ButtonsEnterAnimVariants[i],
      animate: {
        ...ButtonsEnterAnimVariants[i].animate,
        y: `${-i * 3.5 - 3}em`,
      },
    };
  }

  return (
    <ButtonContainer data-testid="mode-buttons">
      <FirstModeButton
        variant="contained"
        component={motion.div}
        exit={ButtonExitAnim}
        onClick={() => changeBreathMode('heartCoherence')}
      >
        {BreathModesDisplay.heartCoherence}
      </FirstModeButton>
      <ModeButton
        variant="contained"
        component={motion.div}
        exit={ButtonExitAnim}
        onClick={() => changeBreathMode('vitality')}
      >
        {BreathModesDisplay.vitality}
      </ModeButton>
      <ModeButton
        variant="contained"
        component={motion.div}
        exit={ButtonExitAnim}
        onClick={() => changeBreathMode('relaxation')}
      >
        {BreathModesDisplay.relaxation}
      </ModeButton>
      <ModeButton
        variant="contained"
        component={motion.div}
        exit={ButtonExitAnim}
        onClick={() => changeBreathMode('square')}
      >
        {BreathModesDisplay.square}
      </ModeButton>
      <ModeButton
        variant="contained"
        component={motion.div}
        exit={ButtonExitAnim}
        onClick={() => changeBreathMode('custom')}
      >
        {BreathModesDisplay.custom}
      </ModeButton>
    </ButtonContainer>
  );
};

const ButtonContainer = styled(motion.div)({
  display: 'flex',
  flexDirection: 'column-reverse',
});

export type ExitAnim = {
  y?: string;
  transition: { duration: number; delay?: number; ease?: string };
};
