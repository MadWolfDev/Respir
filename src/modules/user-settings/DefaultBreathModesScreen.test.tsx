import { render } from '@testing-library/react';
import { DefaultBreathModesScreen } from './DefaultBreathModesScreen';

describe('DefaultBreathModesScreen', () => {
  it('should display all the buttons', () => {
    const component = render(<DefaultBreathModesScreen />);
    expect(component.baseElement).toMatchSnapshot();
  });
});
