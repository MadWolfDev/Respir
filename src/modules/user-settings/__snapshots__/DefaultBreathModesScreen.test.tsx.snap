// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DefaultBreathModesScreen should display all the buttons 1`] = `
<body>
  <div>
    <div
      class="css-1p70m1z"
    >
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-gk8r2c-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="button"
      >
        Cohérence cardiaque
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-15l7ny6-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="button"
      >
        Respiration vitalité
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-15l7ny6-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="button"
      >
        Respiration détente
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-15l7ny6-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="button"
      >
        Respiration carrée
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-15l7ny6-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="button"
      >
        Personnalisé
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</body>
`;
